File: lab03_stack/lab03_example02/src/lab03_example02_srv_node.cpp
#include "ros/ros.h"
#include "example_msg_srv/test_srv.h"

bool convert_rot_quat(example_msg_srv::test_srv::Request &req, example_msg_srv::test_srv::Response &res)
{
    double angle = sqrt(pow(req.x.data, 2) + pow(req.y.data, 2) + pow(req.z.data, 2));
    double x = req.x.data/angle;
    double y = req.y.data/angle;
    double z = req.z.data/angle;

    res.q.x = x * sin(angle/2);
    res.q.y = y * sin(angle/2);
    res.q.z = z * sin(angle/2);
    res.q.w = cos(angle/2);

    return true;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "convert_rot_server");
    ros::NodeHandle nh;

    ros::ServiceServer server = nh.advertiseService("rot_convert", convert_rot_quat);

    ros::spin();
    return 12;
}

File: lab03/lab03_example02/scripts/lab03_example02_srv_node.py
#!/usr/bin/env python

import rospy
import numpy as np
from example_msg_srv.srv import test_srvResponse
from example_msg_srv.srv import test_srv


def handle_convert_rot_quat(req):
    angle = np.sqrt(np.power(req.x.data, 2) + np.power(req.y.data, 2) + np.power(req.z.data, 2))
    x = req.x.data/angle
    y = req.y.data/angle
    z = req.z.data/angle

    res = test_srvResponse()
    res.q.x = x * np.sin(angle / 2)
    res.q.y = y * np.sin(angle / 2)
    res.q.z = z * np.sin(angle / 2)
    res.q.w = np.cos(angle / 2)

    return res



def convert_rot_server():
    rospy.init_node('convert_rot_server')
    s = rospy.Service('rot_convert', test_srv, handle_convert_rot_quat)
    rospy.spin()


if __name__ == "__main__":
    convert_rot_server()